arangod/Agency/Agent.cpp

```diff
@@ -1298,6 +1415,54 @@ read_ret_t Agent::read(query_t const& query) {
   return read_ret_t(true, leader, std::move(success), std::move(result));
 }
 
+
+// trigger all polls, who have timed out with empty result
+void Agent::clearExpiredPolls() {
+  index_t commitIndex = 0;
+  {
+    READ_LOCKER(oLocker, _outputLock);
+    commitIndex = _commitIndex;
+  }
+  auto empty = std::make_shared<VPackBuilder>();
+  {
+    VPackObjectBuilder obj(empty.get());
+    empty->add("accepted", VPackValue(true));
+    empty->add(VPackValue("result"));
+    VPackObjectBuilder res(empty.get());
+    empty->add("firstIndex", VPackValue(commitIndex));
+    empty->add("commitIndex", VPackValue(commitIndex));
+    empty->add(VPackValue("log"));
+    VPackArrayBuilder a(empty.get());
+  }
+  std::lock_guard lck(_promLock);
+  triggerPollsNoLock(empty, std::chrono::steady_clock::now());
+}
+
+
+/// Clear expired polls
+/// Wake up everybody with quertand delete with empty
+/// If qu is nullptr, we're resigning.
+void Agent::triggerPollsNoLock(query_t qu, SteadyTimePoint const& tp) {
+
+  if (qu == nullptr) { // We have resigned
+    qu = std::make_shared<VPackBuilder>();
+    VPackObjectBuilder qb(qu.get());
+    qu->add("accepted", VPackValue(false));
+    qu->add(VPackValue("result"));
+    VPackArrayBuilder arr(qu.get());
+  }
+
+  auto it = _promises.begin();
+  while (it != _promises.end()) {
```

Comment:
The common case will be that all promises are woken up. We could optimize for this by not erasing the promises one by one but by erasing the whole map in the end. Also in the other case we could erase a range, which might be more efficient in some implementations...

Reply:
very good point. awesome suggestion.